[HLSL] Implement the `degrees` intrinsic

inbelic opened the pull request:
- add degrees builtin
    - link degrees api in hlsl_intrinsics.h
    - add degrees intrinsic to IntrinsicsDirectX.td
    - add degrees intrinsic to IntrinsicsSPIRV.td
    - add lowering from clang builtin to dx/spv intrinsics in CGBuiltin.cpp
    - add semantic checks to SemaHLSL.cpp
    - add expansion of directx intrinsic to llvm fmul for DirectX in DXILIntrinsicExpansion.cpp
    - add mapping to spir-v intrinsic in SPIRVInstructionSelector.cpp

    - add test coverage:
        - degrees.hlsl -> check hlsl lowering to dx/spv degrees intrinsics
        - degrees-errors.hlsl/half-float-only-errors -> check semantic warnings
        - hlsl-intrinsics/degrees.ll -> check lowering of spir-v degrees intrinsic to SPIR-V backend
        - DirectX/degrees.ll -> check expansion and scalarization of directx degrees intrinsic to fmul
      
Resolves #99104

inbelic committed:
[HLSL] Implementation the `degrees` intrinsic

    - add degrees builtin
    - link degrees api in hlsl_intrinsics.h
    - add degrees intrinsic to IntrinsicDirectX.td
    - add degrees intrinsic to IntrinsicSPIRV.td
    - add lowering from clang builtin to dx/spv intrinsics in
      CGBuiltin.cpp
    - add semantic checks to SemaHLSL.cpp
    - add expansion of directx intrinsic to llvm fmul for DirectX in
      DXILIntrinsicExpansion.cpp
    - add mapping to spir-v intrinsic in SPIRVInstructionSelector.cpp

    - add test coverage:
        - degrees.hlsl -> check hlsl lowering to dx/spv degrees intrinsics
        - degrees-errors.hlsl/half-float-only-errors ->
            check semantic warnings
        - hlsl-intrinsics/degrees.ll -> check lowering of spir-v
            degrees intrinsic to SPIR-V backend
        - DirectX/degrees.ll -> check expansion and scalarization of
          directx degrees intrinsic to fmul

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -1643,6 +1646,23 @@ bool SPIRVInstructionSelector::selectLength(Register ResVReg,
+                                             const SPIRVType *ResType,
+                                             MachineInstr &I) const {
+
+  assert(I.getNumOperands() == 3);
```
Doesn't degrees have 1 operand?

inbelic replied:
Correct, but I believe the first 2 operands of the spir-v instruction are usually used for the `ReturnType` and the `Result` register.

spall commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2625,6 +2645,8 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_length:
     return selectLength(ResVReg, ResType, I);
+  case Intrinsic::spv_degrees:
+    return selectDegrees(ResVReg, ResType, I);
```
Assuming it does take 1 operand, can you use selectExtInst here instead of a custom function?

inbelic replied:
Yep, you are right and I now realize I should also support the option to use the opencl lowering path which can be done using `selectExtInst`.

inbelic commented:
Rebasing onto commit that allows the use of `selectExtInst` with both `CL::degrees` and `GL::Degrees`. Adding additional testcase for this as well.

pow2clk commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/degrees.ll:
```diff
@@ -0,0 +1,68 @@
+; CHECK-DAG: %[[#vec4_float_64:]] = OpTypeVector %[[#float_64]] 4
+
+define noundef float @degrees_float(float noundef %a) {
+entry:
```
It's not essential, but separating these with some `CHECK-LABEL`s for the function definitions even if they are nothing more than `; CHECK: define{{.*}} degrees_float` would help with potential future failures particularly if the only initial information is an automated testing failure.

pow2clk reviewed (commented):
Looks great overall!

None of my suggestions are essential, but I'll give you a chance to consider them even if you reject them all before I approve.

inbelic committed:
review comments:

improve testcases:
    - remove flags that were not needed
    - improve readability of FileCheck variables
    - add CHECK-LABEL to spirv check

pow2clk commented on llvm/test/CodeGen/SPIRV/opencl/degrees.ll:
```diff
@@ -0,0 +1,50 @@
+; CHECK: %[[#]] = OpExtInst %[[#vec4_float_16]] %[[#op_ext_glsl]] degrees %[[#vec4_float_16_arg]]
+  %elt.degrees = call <4 x half> @llvm.spv.degrees.v4f16(<4 x half> %a)
+  ret <4 x half> %elt.degrees
+}
```
Per conversation with @farzonl, I'm fine with accepting, but not requiring such tests.

pow2clk reviewed (approved):
Thanks! Though none of these were essential, I think it makes for a better change.

inbelic committed:
Merge branch 'main' into inbelic/hlsl-degrees

inbelic committed:
clang format

inbelic committed:
self review:

    - remove bad test case as double is not supported for hlsl or spirv
      backend
